packages/loader/container-loader/src/protocol.ts

```diff
@@ -75,11 +75,12 @@ export class ProtocolHandler extends ProtocolOpHandler implements IProtocolHandl
 		message: ISequencedDocumentMessage,
 		local: boolean,
 	): IProcessMessageResult {
-		const client: ISequencedClient | undefined = this.quorum.getMember(message.clientId);
+		const client: ISequencedClient | undefined =
+			message.clientId === null ? undefined : this.quorum.getMember(message.clientId);
 
 		// Check and report if we're getting messages from a clientId that we previously
 		// flagged as shouldHaveLeft, or from a client that's not in the quorum but should be
-		if (message.clientId != null) {
+		if (message.clientId !== null) {
```

Comment:
This seems like a mistake, so corrected it.

Reply:
I thought `!== null` was slower than `!= null` without any advantage. Can't we rely on linter to catch the problem cases?

Reply:
Strict equality is best practice here.  There are certain places in the repo that intentionally distinguish `undefined` from `null`, so it's better to be consistent throughout.  Also in general our performance bottlenecks aren't going to be on comparisons - if we ever profile a scenario and discover that to be the case we could have the conversation.

Reply:
And I had the perf wrong - it so often doesn't come up and I flipped.
There is a difference here for treatment of `message.clientId` being `undefined`. The prior code did a strict undefined check on `client` afterward which was `message.clientId` passed through `this.quorum.getMember` (which I don't yet know anything about).
In the latest version of changes there are checks that don't make sense. As of commit 189f54d58
```typescript
		if (message.clientId !== null) {
			const client =
				message.clientId === null ? undefined : this.quorum.getMember(message.clientId);
```
The inner `message.clientId === null` will never be true. Or am I crossing other wires?

If it is possible to make these changes separately, that would be good.

Reply:
Hm yeah maybe just defer this fix for now - I'm not entirely sure what it should be doing here so probably needs some investigation (join ops have a null clientId, so maybe the first error is also unreachable?)

Reply:
I can't just leave it as it was, because `null` is not a valid value to pass to getMember. I have to handle the null
case explicitly. I changed the client declaration to be
```ts
const client =
			message.clientId === null ? undefined : this.quorum.getMember(message.clientId);
```

I think that's equivalent assuming `this.quorum.getMember(null)` returns undefined.

Reply:
Moving the call inside if should do the trick, yeah? No other change needed.
(I am assuming getMember has no desired side-effects.)

Reply:
I'm a little confused. In an earlier iteration, the code was this:

```ts
		// Check and report if we're getting messages from a clientId that we previously
		// flagged as shouldHaveLeft, or from a client that's not in the quorum but should be
		if (message.clientId !== null) {
			const client =
				message.clientId === null ? undefined : this.quorum.getMember(message.clientId);

			if (client === undefined && message.type !== MessageType.ClientJoin) {
```

But that has the problem you noted where `message.clientId` can never be null inside the if so the other if is always skipped.

Now it's:

```ts
		const client =
			message.clientId === null ? undefined : this.quorum.getMember(message.clientId);

		// Check and report if we're getting messages from a clientId that we previously
		// flagged as shouldHaveLeft, or from a client that's not in the quorum but should be
		if (message.clientId !== null) {
			if (client === undefined && message.type !== MessageType.ClientJoin) {
```

I think you're suggesting:

```ts
		// Check and report if we're getting messages from a clientId that we previously
		// flagged as shouldHaveLeft, or from a client that's not in the quorum but should be
		if (message.clientId !== null) {
			const client = this.quorum.getMember(message.clientId);
			if (client === undefined && message.type !== MessageType.ClientJoin) {
```

Did I understand correctly?

Reply:
Close. I am suggesting not to mess with `if (message.clientId != null)` now. So
``typescript
// Check and report if we're getting messages from a clientId that we previously
// flagged as shouldHaveLeft, or from a client that's not in the quorum but should be
if (message.clientId != null) {
    const client = this.quorum.getMember(message.clientId);
```
If you change the `if` to strict then when clientId is undefined, it will go into the block. Previously it would not.

Reply:
> If you change the if to strict then when clientId is undefined, it will go into the block. Previously it would not.

Ahhhh, thank you. Makes sense to me now. I find null vs undefined and strict vs non-strict difficult to reason about.

Updated back to `!=`.